src/jvm/src/aws/sdk/kotlin/crt/auth/credentials/StsAssumeRoleCredentialsProviderJVM.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.sdk.kotlin.crt.auth.credentials
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.future.asCompletableFuture
+import kotlinx.coroutines.runBlocking
+import java.util.concurrent.CompletableFuture
+import software.amazon.awssdk.crt.auth.credentials.Credentials as CredentialsJni
+import software.amazon.awssdk.crt.auth.credentials.CredentialsProvider as CredentialsProviderJni
+import software.amazon.awssdk.crt.auth.credentials.StsCredentialsProvider as StsCredentialsProviderJni
+
+// Based from the Java SDK default
+// https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/sts/model/AssumeRoleRequest.html#durationSeconds
+private const val DEFAULT_DURATION_SECONDS = 3600
+
+public actual class StsAssumeRoleCredentialsProvider
+internal actual constructor(builder: StsAssumeRoleCredentialsProviderBuilder) :
+    CredentialsProvider, JniCredentialsProvider() {
+    public actual companion object {}
+
+    override val jniCredentials: CredentialsProviderJni =
+        StsCredentialsProviderJni
+            .builder()
+            .withClientBootstrap(builder.clientBootstrap?.jniBootstrap)
+            .withTlsContext(builder.tlsContext?.jniCtx)
+            .withCredsProvider(adapt(builder.credentialsProvider))
+            .withDurationSeconds(builder.durationSeconds ?: DEFAULT_DURATION_SECONDS)
+            .withRoleArn(builder.roleArn)
+            .withSessionName(builder.sessionName)
+            .build()
+}
+
+// Convert the SDK version of CredentialsProvider type to the CRT version
+internal fun adapt(credentialsProvider: CredentialsProvider?): CredentialsProviderJni? {
```

Comment:
suggestion

I was going to ask about this. I'm assuming this is because the `CredentialsProvider` used for sourcing credentials *MUST* be one of the `aws-crt-java` credential providers to actually work (see https://github.com/awslabs/aws-crt-kotlin/issues/15). An arbitrary customer provided provider will fail. 

Instead of using `runBlocking` we could just do this at the time `getCredentials` is called. So instead of inheriting from `JniCredentialsProvider()`, just specialize this one since it has custom requirements.

e.g. something like

```kotlin

public actual class StsAssumeRoleCredentialsProvider 
    internal actual constructor(builder: Builder) : CredentialsProvider {
     
     private val sourceCredentialsProvider = builder.credentialsProvider
     
     private val jniBuilder = StsCredentialsProviderJni.builder()
         .withClientBootstrap(...)
         // ... everything except withCredsProvider()
         // leave it as a builder() not fully constructed

      override suspend fun getCredentials(): Credentials {
           val sourceCreds = sourceCredentialsProvider.getCredentials()
           val staticProvider = StaticCredentialsProviderJni(sourceCreds)
           
           // complete building the JNI provider here, it will be a single use provider
           val stsProvider = jniBuilder.withCredsProvider(staticProvider)
           val jniCreds = stsProvider.use {
               stsProvider.credentials.await()                    
           }

        return Credentials(jniCreds.accessKeyId, jniCreds.secretAccessKey, jniCreds.sessionToken)
      }

    // we keep no references to an actual provider past getCredentials() so close/waitForShutdown are no-ops
    override fun close() {}
    override suspend fun waitForShutdown() { }
}
```

Reply:
Tracking issue: https://github.com/awslabs/aws-sdk-kotlin/issues/357